[Refactor] Use iterative approach to evaluate Regex.simpleMatch

cwperks opened the pull request:
### Description

In the internal OpenSearch class for [Regex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java), its possible to supply patterns like `abc**` which contain adjacent wildcards. When processing inputs with adjacent wildcards, the internal method relies on [recursion](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L131-L161) and is performing unnecessary recursive calls when handling inputs with adjacent wildcards which are redundant in a regex. 

This PR contains an optimization to the way that the internal regex evaluates a simple match by eliminating the need to recurse when an input contains adjacent wildcards.

----

Additional context:

When trying to compile a `java.regex.Pattern` similarly like this `Pattern.compile("abc**");`, it fails with the following error:

```
java.util.regex.PatternSyntaxException: Dangling meta character '*' near index 4
```

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

cwperks commented:
Closing while additional test cases are added.

reta commented:
@cwperks my apologies if I am missing something, but why we should alter the pattern if it is not supplied correctly? The current behaviour that raises an exception is looking legitimate to me.

cwperks commented:
@reta It does not raise an exception currently. In the current form it [recurses](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L131-L161) through the input and permits adjacent wildcards. This PR maintains existing behavior, but provides an optimization that will eliminate the need to recurse.

reta commented:
> This PR maintains existing behavior, but provides an optimization that will eliminate the need to recurse.

Got it, thanks @cwperks , the description does not much the change?

cwperks commented:
Apologies @reta, I updated the description to be clearer. I had some additional context in the PR description around how  java.regex.Pattern handled cases with adjacent wildcards and I moved that to another section to make it clearer that that's contextual information.

reta commented:
> Apologies @reta, I updated the description to be clearer. I had some additional context in the PR description around how java.regex.Pattern handled cases with adjacent wildcards and I moved that to another section to make it clearer that that's contextual information.

Gotcha, thanks a lot @cwperks , I think I understood the problem clearly now

reta commented on server/src/main/java/org/opensearch/common/regex/Regex.java:
```diff
@@ -93,6 +93,23 @@ public static Automaton simpleMatchToAutomaton(String... patterns) {
+     * @param target - The target character to remove duplicates of
+     * @return - string with adjacent duplicates of the target character removed
+     */
+    static String removeAdjacentDuplicates(String str, char target) {
```
Won't the simple change in `simpleMatchWithNormalizedStrings` fix the problem by skipping over the `*` sequence? 

```
            } else if (nextIndex == 1) {
                // The wildcard series, "**" - skipping all till the last "*"
                int skipTillIndex = nextIndex + 1;
                while (pattern.length() < skipTillIndex && pattern.charAt(skipTillIndex) == '*') { 
                    ++skipTillIndex;
                };
                return simpleMatchWithNormalizedStrings(pattern.substring(skipTillIndex - 1), str);
            }
```

cwperks replied:
Either approach is fine. I chose this approach aiming to ensure that the input to the `simpleMatchWithNormalizedStrings` was valid so that the logic for this edge case did not need to reside within the method. It could be considered part of the `normalization` of the input.

reta replied:
The `simpleMatchWithNormalizedStrings` for some reasons accepted only `**` but not the sequences of it, modify the regex in this cases basically is possible but there is clearly more than one place now that deals with `*`, consolidating the logic in a single place seems to make sense.

cwperks replied:
This PR performs the [normalization](https://github.com/opensearch-project/OpenSearch/blob/57ff3cdf6ae186218983df6bd3a79e4a3f8bcd55/server/src/main/java/org/opensearch/common/regex/Regex.java#L138-L145) before feeding the normalized input into `simpleMatchWithNormalizedStrings`. There is currently only one place where `simpleMatchWithNormalizedStrings` is called. I chose to put the logic before the invocation of `simpleMatchWithNormalizedStrings` so that its only called once before and not on each recurse of the call. 

Do you think it makes sense to put the call elsewhere to normalize the input?

reta replied:
> Do you think it makes sense to put the call elsewhere to normalize the input?

I see 2 issues with normalization approach:
- if you look for the exception that `simpleMatchWithNormalizedStrings` throws now, it prints out the **normalized** pattern and not what user has really sent out (`trimmedPattern` vs `pattern`), we are losing the truth source here
- we do double work since we still have to go and parse whole pattern (but we do this twice now, one time to fold the `*` and second time to actually go and parse it again)

It seemed to me that folding `*` in one go is simpler and more straightforward approach, and it does not require pre-normalization call.

DarshitChanpura replied:
> if you look for the exception that simpleMatchWithNormalizedStrings throws now, it prints out the normalized pattern and not what user has really sent out (trimmedPattern vs pattern), we are losing the truth source here

This seems solvable by passing original pattern

>we do double work since we still have to go and parse whole pattern (but we do this twice now, one time to fold the * and second time to actually go and parse it again)

This would add a complexity of m making it O(n+m) with n>=m, where in worst case scenario where there are no `*`s, m= n leading to O(2n) which would still be constant time.

> It seemed to me that folding * in one go is simpler and more straightforward approach, and it does not require pre-normalization call.

This could be followed if its guaranteed that `simpleMatchWithNormalizedStrings` will be called only once

dbwiddis replied:
The mere presence of the "simple match" optimization indicates it's intended for performance-sensitive/hot-path applications to shortcut regex.

I had originally started a comment saying that this approach was going to create another String just to temporarily toss it onto the garbage heap, but then I realized the entire method is recursive for every wildcard (even non-adjacent ones).    So the String "opensearch*is*an*awesome*search*engine" will recurse 5 times, generating more trash to be GC'd.  If you're taking the effort to eliminate one recursion in "**" let's go all the way and eliminate all the recursion.

Your one-pass method here is a step in the right direction but you can do even better by doing a single pass over each of the two strings without ever creating any new String objects.

Suggestion:
 - Keep the original `pattern` and `str` objects, never create new ones (or substrings)
 - Keep an `int` starting index pointer for each of the two strings
 - Use `indexOf()` to find the wildcard position and `regionMatches()` to compare up until that point 
 - Use `indexOf()` overload to find the next index (replace the if/then checking only one dupe with a loop until wildcards are done as @reta suggested)
 - Use `substring()` to find the bits in between these indices (current temporary `part` String, don't think there's an easy way around that) and the position of that in str
 - repeat

That suggestion mostly follows the existing code except for adding @reta's improvement and replacing the final recursive line with a loop construct where you use the index pointers instead of 0.

 - 
1. Don't use `str.toCharArray()` in your iteration, that's another creation of a new (array) object just to throw it away.  Keep the original strings and just iterate using an integer and `charAt()`.
2. With a com

dbwiddis replied:
OK, so I have just spent most of the last hour trying to implement the approach I outlined above.  And it just ain't worth the sacrifice of readability for the slight improvement in efficiency.

I did, in my various iterations, come upon what I think is an elegant solution which is much more readable than the current recursive mess:

1. Alter this method to return a `List<String>` instead of just a String.  Do this by:
   - iterate by character and adding to a StringBuilder (except use `charAt()` other than an array)
   - when you encounter a wildcard, add the existing string to the list if it's not empty, and reset the builder
2. In the `simpleMatch()` method just track a "position" in the string, starting at 0
3. Iterate your list and find str.indexOf(substringfromlist, startpos) advancing the str pointer each time.  Pos should start at exactly 0 (unless special case of pattern starting with *) otherwise >0.
4. Advance to that index + length.   Repeat iteration from 3.
5. Special case the ends-with wildcard.

cwperks replied:
Thank you @dbwiddis. I updated the algorithm to remove recursion entirely. It now uses an iterative 2 pointer-style solution.

dbwiddis reviewed (commented):
While this fix simply removes the adjacent duplicates, if the goal is reducing recursion you can do much better

scrawfor99 commented:
@reta @dbwiddis given what Dan found and the changes that Craig made, I am hoping we can move forward with this as is. 

I think that it solves the primary issue of the adjacent wildcards while also avoiding a lot of the overhead and intricacies that Dan was running into. 

Does that sound reasonable?

reta commented:
> @reta @dbwiddis given what Dan found and the changes that Craig made, I am hoping we can move forward with this as is.

The algorithm changes look worth it, :+1: to keep it

reta commented on server/src/test/java/org/opensearch/common/regex/RegexTests.java:
```diff
@@ -96,6 +96,7 @@ public void testDoubleWildcardMatch() {
         assertTrue(Regex.simpleMatch("fff*******ddd", "fffabcddd"));
         assertTrue(Regex.simpleMatch("fff*******ddd", "FffAbcdDd", true));
+        assertFalse(Regex.simpleMatch("fff*******ddd", "FffAbcdDd", false));
         assertFalse(Regex.simpleMatch("fff******ddd", "fffabcdd"));
```
Could we add please more "bogus" tests:

```
        assertTrue(Regex.simpleMatch("*******************************************", "fffabcddd"));
        assertTrue(Regex.simpleMatch("************f*************b*******d***********", "fffabcddd"));
        assertTrue(Regex.simpleMatch("fff*******dd*****", "fffabcddd"));
        assertTrue(Regex.simpleMatch("****ff*******dd*****", "fffabcddd"));

```

scrawfor99 replied:
Yeah will do

dbwiddis replied:
Need to make sure we cover all possible edge cases in the tests in both single and multiple wildcard cases
 - wildcard at start
 - wildcard at end
 - wildcard both at start and end
 - 0 wildcards
 - 1 wildcard
 - more than 1 wildcard

dbwiddis reviewed (approved):
LGTM, pending additional test cases as requested!

scrawfor99 commented:
Added!

scrawfor99 commented on CHANGELOG.md:
```diff
@@ -63,6 +63,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Return 409 Conflict HTTP status instead of 503 on failure to concurrently execute snapshots ([#8986](https://github.com/opensearch-project/OpenSearch/pull/5855))
 - Add task completion count in search backpressure stats API ([#10028](https://github.com/opensearch-project/OpenSearch/pull/10028/))
 - Performance improvement for Datetime field caching ([#4558](https://github.com/opensearch-project/OpenSearch/issues/4558))
+- Use iterative approach to evaluate Regex.simpleMatch ([#11060](https://github.com/opensearch-project/OpenSearch/pull/11060))
```
```suggestion
```

scrawfor99 commented on CHANGELOG.md:
```diff
@@ -129,6 +129,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - [BUG] Disable sort optimization for HALF_FLOAT ([#10999](https://github.com/opensearch-project/OpenSearch/pull/10999))
 - Performance improvement for MultiTerm Queries on Keyword fields ([#7057](https://github.com/opensearch-project/OpenSearch/issues/7057))
 - Disable concurrent aggs for Diversified Sampler and Sampler aggs ([#11087](https://github.com/opensearch-project/OpenSearch/issues/11087))
+- Remove adjacent duplicate wildcard in Regex ([#11060](https://github.com/opensearch-project/OpenSearch/pull/11060))
```
```suggestion
- Use iterative approach to evaluate Regex.simpleMatch ([#11060](https://github.com/opensearch-project/OpenSearch/pull/11060))
```

scrawfor99 commented:
This seems unrelated: 

```

> Task :qa:wildfly:composeUp FAILED
received unexpected HTTP status: 500 Internal Server Error
```

cwperks commented:
It took a few runs due to flaky tests, but the CI is finally green. @reta or @dbwiddis could you please add the backport 2.x label and merge this PR?

scrawfor99 commented:
@reta, can you try a backport to 1.x? I saw that 2.x failed but if 1.x also fails I will manually backport
